server/models/annotation.py

```diff
@@ -555,16 +586,20 @@ def remove(self, annotation, *args, **kwargs):
 
         :param annotation: the annotation document to remove.
         """
-        delete_one = self.collection.delete_one
+        if self._historyEnabled:
+            # just mark the annotations as inactive
+            result = self.update({'_id': annotation['_id']}, {'$set': {'_active': False}})
+        else:
+            delete_one = self.collection.delete_one
 
-        def deleteElements(query, *args, **kwargs):
-            ret = delete_one(query, *args, **kwargs)
-            Annotationelement().removeElements(annotation)
-            return ret
+            def deleteElements(query, *args, **kwargs):
+                ret = delete_one(query, *args, **kwargs)
+                Annotationelement().removeElements(annotation)
+                return ret
 
-        self.collection.delete_one = deleteElements
-        result = super(Annotation, self).remove(annotation, *args, **kwargs)
-        self.collection.delete_one = delete_one
+            self.collection.delete_one = deleteElements
```

Comment:
Overriding methods on the mongo collection in a rest request makes me very uncomfortable.  Doesn't this create some pretty ugly race conditions?

Given that this is done elsewhere in the Annotation model, maybe this PR isn't the correct place to solve it, but I would like to have us explore an alternative mechanism for dealing with the interaction between the Annotation and Element collections.

Reply:
Mongo doesn't have the concept of a transaction lock, nor can I find any mechanism for placing a write lock on a collection.  As long as annotations and annotation elements are in two collections, we have to manually do something that is equivalent to a transaction.

When writing, we get a new version (which is guaranteed to be unique), save the new elements (which have a distinct version), then update the annotation with information that includes the new version.  When saving history, we copy the existing annotation record to a new record (which gets a new ID but keeps the old version.  It is possible if more than one process is updating an annotation, this copying can happen more than once.  There will only be one annotation record with the latest version, but there *could be* more than one copy of the annotation with the same old version number.

In reconsidering this, I do need to add a lock inside the annotation collection so that we can ensure that when we override the methods it is guarded.

I would love to find a better method for mongo transactions.

Reply:
I think we are hitting limitations of mongo here because we are trying to use it in a way that it isn't intended to be used.  Transactions are just not something we should try and implement.  Likely locks are the best immediate solution to this particular problem, but I wouldn't consider this a pattern to encourage elsewhere.  My hope is that in the future, we can do more to design our schemas in a way that these hacks aren't necessary.